gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java

```diff
@@ -0,0 +1,302 @@
+            ExecutorsUtils.newThreadFactory(Optional.of(log), Optional.of(getClass().getSimpleName()))));
+  }
+
+  /**
+   * Register a table or partition given a {@link HiveSpec}.
+   *
+   * This method is asynchronous and returns immediately.
+   *
+   * @return a {@link ListenableFuture} for the process of registering the given {@link HiveSpec}.
+   */
+  public ListenableFuture<Void> register(final HiveSpec spec) throws IOException {
+    ListenableFuture<Void> future = this.executor.submit(new Callable<Void>() {
+
+      @Override
+      public Void call() throws Exception {
+
+        if (spec instanceof HiveSpecWithPredicates && !evaluatePredicates((HiveSpecWithPredicates) spec)) {
+          log.info("Skipping " + spec + " since predicates return false");
+        }
+
+        if (spec instanceof HiveSpecWithPreActivities) {
+          for (Activity<HiveMetaStoreClient> activity : ((HiveSpecWithPreActivities) spec).getActivities()) {
+            activity.execute(HiveRegister.this.client);
+          }
+        }
+
+        String dbName = spec.getDbName();
+        String tableName = spec.getTableName();
+
+        createDbIfNotExists(dbName);
+        Optional<HivePartition> partition = spec.getPartition();
+        Table table = getOrCreateTable(dbName, tableName, partition, spec.getSd());
+
+        if (table.getPartitionKeys().isEmpty()) {
+
+          // Register a non-partitioned table
+          HiveRegister.this.client.alter_table(dbName, tableName, table);
+        } else {
+
+          // Register a partition
+          addPartition(table, partition.get());
+        }
+
+        if (spec instanceof HiveSpecWithPostActivities) {
+          for (Activity<HiveMetaStoreClient> activity : ((HiveSpecWithPostActivities) spec).getActivities()) {
+            activity.execute(HiveRegister.this.client);
+          }
+        }
+
+        return null;
+      }
+    });
+
+    this.futures.add(future);
+    return future;
+  }
+
+  private boolean evaluatePredicates(HiveSpecWithPredicates spec) {
+    for (Predicate<HiveMetaStoreClient> pred : spec.getPredicates()) {
+      if (!pred.apply(this.client)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private void createDbIfNotExists(String dbName) {
+    Database db = new Database();
+    db.setName(dbName);
+
+    Preconditions.checkState(this.hiveDbRootDir.isPresent(),
+        "Missing required property " + HiveRegProps.HIVE_DB_ROOT_DIR);
+    db.setLocationUri(this.hiveDbRootDir.get() + Path.SEPARATOR + dbName + HIVE_DB_EXTENSION);
+
+    Lock lock = locks.get(dbName);
+    lock.lock();
+
+    try {
+      client.createDatabase(db);
+    } catch (AlreadyExistsException e) {
+      log.warn("Hive database " + dbName + " already exists");
+    } catch (TException e) {
+      log.error("Unable to create Hive database " + dbName);
+      throw Throwables.propagate(e);
```

Comment:
This is just in case the exception message is not clear enough. Adding a log here to ensure printing a message that we can control.